libobs/util/buffered-file-serializer.c

```diff
@@ -0,0 +1,415 @@
+				os_event_reset(
+					out->io.new_data_available_event);
+				pthread_mutex_unlock(&out->io.data_mutex);
+				break;
+			}
+
+			pthread_mutex_unlock(&out->io.data_mutex);
+
+			// Seek if we need to
+			if (want_seek) {
+				os_fseeki64(out->io.output_file,
+					    next_seek_position, SEEK_SET);
+
+				// Update the next virtual position, making sure to take
+				// into account the size of the chunk we're about to write.
+				current_seek_position =
+					next_seek_position + chunk_used;
+
+				want_seek = false;
+
+				// If we did a seek but do not have any data left to write
+				// return to the start of the loop.
+				if (!chunk_used) {
+					force_flush_chunk = false;
+					continue;
+				}
+			}
+
+			// Write the current chunk to the output file
+			size_t bytes_written = fwrite(chunk, 1, chunk_used,
+						      out->io.output_file);
+			if (bytes_written != chunk_used) {
+				blog(LOG_ERROR,
+				     "Error writing to '%s': %s (%zu != %zu)\n",
+				     out->filename.array, strerror(errno),
+				     bytes_written, chunk_used);
+				os_atomic_set_bool(&out->io.output_error, true);
+
+				goto error;
+			}
+
+			chunk_used = 0;
+			force_flush_chunk = false;
+		}
+
+		// If this was the last chunk, time to exit
+		if (shutting_down)
+			break;
+	}
+
+error:
+	if (chunk)
+		bfree(chunk);
+
+	fclose(out->io.output_file);
+	return NULL;
+}
+
+/* ========================================================================== */
+/* Serializer Implementation                                                  */
+
+static int64_t file_output_seek(void *opaque, int64_t offset,
+				enum serialize_seek_type seek_type)
+{
+	struct file_output_data *out = opaque;
+
+	// If the output thread failed, signal that back up the stack
+	if (os_atomic_load_bool(&out->io.output_error))
+		return -1;
+
+	// Update where the next write should go
+	pthread_mutex_lock(&out->io.data_mutex);
+
+	switch (seek_type) {
+	case SERIALIZE_SEEK_START:
+		out->io.next_pos = offset;
+		break;
+	case SERIALIZE_SEEK_CURRENT:
+		out->io.next_pos += offset;
+		break;
+	case SERIALIZE_SEEK_END:
+		out->io.next_pos -= offset;
+		break;
+	}
+
+	pthread_mutex_unlock(&out->io.data_mutex);
+
+	return (int64_t)out->io.next_pos;
+}
+
+#ifndef _WIN32
+static inline size_t max(size_t a, size_t b)
+{
+	return a > b ? a : b;
+}
+
+static inline size_t min(size_t a, size_t b)
+{
+	return a < b ? a : b;
+}
+#endif
+
+static size_t file_output_write(void *opaque, const void *buf, size_t buf_size)
+{
+	struct file_output_data *out = opaque;
+
+	if (!buf_size)
+		return 0;
+
+	// Split writes into at chunks that are at most chunk_size bytes
+	uintptr_t ptr = (uintptr_t)buf;
+	size_t remaining = buf_size;
+
+	while (remaining) {
+		if (os_atomic_load_bool(&out->io.output_error))
+			return 0;
+
+		pthread_mutex_lock(&out->io.data_mutex);
+
+		size_t next_chunk_size = min(remaining, out->io.chunk_size);
+
+		// Avoid unbounded growth of the circlebuf, cap to buffer_size
```

Comment:
`data` is a deque. Should the comment reflect that, or is "circlebuf" sufficient here?

Reply:
Yeah the code was directly copied from ffmpeg-mux and I forgot to update some outdated comments.